packages/zoe/src/contracts/vpool-xyk-amm/multipoolMarketMaker.js

```diff
@@ -0,0 +1,193 @@
+
+import '../../../exported.js';
+import { makeMakeCollectFeesInvitation } from '../newSwap/collectFees.js';
+import { makeMakeSwapInvitation } from './swap';
+import { makeDoublePool } from './doublePool';
+
+/**
+ * Multipool AMM is a rewrite of Uniswap that supports multiple liquidity pools,
+ * and direct exchanges across pools. Please see the documentation for more:
+ * https://agoric.com/documentation/zoe/guide/contracts/multipoolAMM.html It
+ * also uses a unique approach to charging fees. Each pool grows on every trade,
+ * and a protocolFee is also extracted.
+ *
+ * We expect that this contract will have tens to hundreds of issuers.  Each
+ * liquidity pool is between the central token and a secondary token. Secondary
+ * tokens can be exchanged with each other, but only through the central
+ * token. For example, if X and Y are two token types and C is the central
+ * token, a swap giving X and wanting Y would first use the pool (X, C) then the
+ * pool (Y, C). There are no liquidity pools directly between two secondary
+ * tokens.
+ *
+ * There should only need to be one instance of this contract, so liquidity can
+ * be shared as much as possible.
+ *
+ * When the contract is instantiated, the central token is specified in the
+ * terms. Separate invitations are available by calling methods on the
+ * publicFacet for adding and removing liquidity and for making trades. Other
+ * publicFacet operations support querying prices and the sizes of pools. New
+ * Pools can be created with addPool().
+ *
+ * When making trades or requesting prices, the caller must specify either a
+ * maximum input amount (swapIn, getInputPrice) or a minimum output amount
+ * (swapOut, getOutPutPrice) or both. For swaps, the required keywords are `In`
+ * for the trader's `give` amount, and `Out` for the trader's `want` amount.
+ * getInputPrice and getOutputPrice each take two Amounts. The price functions
+ * return both amountIn (which may be lower than the original amount) and
+ * amountOut (which may be higher). When both prices are specified, no swap will
+ * be made (and no price provided) if both restrictions can't be honored.
+ *
+ * When adding and removing liquidity, the keywords are Central, Secondary, and
+ * Liquidity. adding liquidity has Central and Secondary in the `give` section,
+ * while removing liquidity has `want` and `give` swapped.
+ *
+ * Transactions that don't require an invitation include addPool and the
+ * queries: getInputPrice, getOutputPrice, getPoolAllocation,
+ * getLiquidityIssuer, and getLiquiditySupply.
+ *
+ * @type {ContractStartFn}
+ */
+const start = zcf => {
+  // This contract must have a "Central" keyword and issuer in the
+  // IssuerKeywordRecord.
+  // poolFee is the portion of the fees that go into the pool (in BP).
+  // protocolFee is the portion of the fees that are shared with validators
+  const {
+    brands: { Central: centralBrand },
+    timer,
+    poolFee,
+    protocolFee,
+  } = zcf.getTerms();
```

Comment:
This comment is mostly a type declaration written in prose. I locally re-wrote it in JSDoc / typescript to confirm my understanding of what the variables are bound to. Perhaps the re-write is worth keeping? Feel free to move `AMMTerms` and/or `BasisPoints` to `types.js`.

Note that the `()` around `(zcf.getTerms())` are part of the typescript cast syntax:

```suggestion
  /**
   * This contract must have a "Central" keyword and issuer in the
   * IssuerKeywordRecord.
   *
   * @typedef {{
   *   brands: { Central: Brand },
   *   timer: TimerService,
   *   poolFee: BasisPoints, // portion of the fees that go into the pool
   *   protocolFee: BasisPoints, // portion of the fees that are shared with validators
   * }} AMMTerms
   *
   * @typedef { bigint } BasisPoints -- hundredths of a percent
   */
  const {
    brands: { Central: centralBrand },
    timer,
    poolFee,
    protocolFee,
  } = /** @type { Terms & AMMTerms } */ (zcf.getTerms());
```

Reply:
WebStorm doesn't like the internal comments, but tsc doesn't mind it, and apparently it made VSCode happy, so I'll leave them there.